pkg/security/certManager/general/certManager.go

```diff
@@ -130,124 +143,191 @@ func (a *CertManager) Close() {
 		return
 	}
 	for _, ca := range a.config.CAPool {
-		if err := a.fileWatcher.Remove(filepath.Dir(ca)); err != nil {
-			a.logger.Errorf("cannot remove fileWatcher for CAPool directory(%v): %w", filepath.Dir(ca), err)
+		if err := a.fileWatcher.Remove(ca); err != nil {
+			a.logger.Errorf("cannot remove fileWatcher for CAPool(%v): %w", ca, err)
 		}
 	}
 	if a.config.CertFile != "" {
-		if err := a.fileWatcher.Remove(filepath.Dir(a.config.CertFile)); err != nil {
-			a.logger.Errorf("cannot remove fileWatcher for CertFile directory(%v): %w", filepath.Dir(a.config.CertFile), err)
+		if err := a.fileWatcher.Remove(a.config.CertFile); err != nil {
+			a.logger.Errorf("cannot remove fileWatcher for CertFile(%v): %w", a.config.CertFile, err)
 		}
 	}
 	if a.config.KeyFile != "" {
-		if err := a.fileWatcher.Remove(filepath.Dir(a.config.KeyFile)); err != nil {
-			a.logger.Errorf("cannot remove fileWatcher for KeyFile directory(%v): %w", filepath.Dir(a.config.KeyFile), err)
+		if err := a.fileWatcher.Remove(a.config.KeyFile); err != nil {
+			a.logger.Errorf("cannot remove fileWatcher for KeyFile(%v): %w", a.config.KeyFile, err)
 		}
 	}
 	a.fileWatcher.RemoveOnEventHandler(&a.onFileChangeFunc)
 }
 
-func (a *CertManager) getServerCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+func (a *CertManager) getTLSKeyPair() (*tls.Certificate, error) {
 	a.private.mutex.Lock()
 	defer a.private.mutex.Unlock()
-	return &a.private.tlsKeyPair, nil
+	if !a.private.tlsCertNotAfter.IsZero() && time.Now().After(a.private.tlsCertNotAfter) {
+		// current certificate is invalid - force reload
+		_, err := a.loadCertsLocked()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if a.private.tlsKeyPair == nil {
+		return nil, fmt.Errorf("certificate is not loaded")
+	}
+
+	return a.private.tlsKeyPair, nil
```

Comment:
Is `a.private.mutex` just protecting the pointer assignment or should it protect the whole `a.private.tlsKeyPair` object? Because when two goroutines call getTLSKeyPair() they get the same pointer outside the lock. If then one changes the object we have a data race.

Reply:
just protecting `a.private.tlsKeyPair` pointer and data in `a.private.tlsKeyPair` is immutable